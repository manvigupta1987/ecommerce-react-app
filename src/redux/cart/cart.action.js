import CartActionType from './cart.types'

export const toggleCartHidden = () => ({
  type: CartActionType.TOGGLE_CART_HIDDEN
})

export const addItems = (item) => ({
  type: CartActionType.ADD_ITEM,
  payload: item
})

export const removeItems = (item) => ({
  type: CartActionType.REMOVE_ITEM,
  payload: item
})

export const removeSingleItem = (item) => ({
  type: CartActionType.REMOVE_SINGLE_ITEM,
  payload: item
})