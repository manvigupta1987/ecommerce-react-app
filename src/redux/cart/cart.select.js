import {createSelector} from 'reselect';

//input selector --> takes complete state and returns piece of the state. 
const selectCart = state => state.cart;

//createSelector takes two argument : first an array of the input selectors, 
// second: an function that would return the value that you want from this selector. This function
// takes the input return from the inputSelector (first parameter)
export const selectCartItems = createSelector(
  [selectCart],
  cart => cart.cartItems
)

export const selectCartItemsCount = createSelector(
  [selectCartItems],
  cartItems => cartItems.reduce((acc, cartItem) => acc + cartItem.quantity , 0) 
)