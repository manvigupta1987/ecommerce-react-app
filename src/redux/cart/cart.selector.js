import {createSelector} from 'reselect';

//input selector --> takes complete state and returns piece of the state. 
const selectCart = state => state.cart;

//createSelector takes two argument : first an array of the input selectors, 
// second: an function that would return the value that you want from this selector. This function
// takes the input return from the inputSelector (first parameter)
export const selectCartItems = createSelector(
  [selectCart],
  cart => cart.cartItems
)

export const selectCartHidden = createSelector(
  [selectCart],
  cart => cart.hidden
)

export const selectCartItemsCount = createSelector(
  [selectCartItems],
  cartItems => cartItems.reduce((acc, cartItem) => acc + cartItem.quantity , 0) 
)

export const selectCartTotalPrice = createSelector(
  [selectCartItems],
  cartItems => cartItems.reduce((acc, cartItem) => acc + cartItem.price * cartItem.quantity, 0)
)