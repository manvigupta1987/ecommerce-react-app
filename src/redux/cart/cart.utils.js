export const addItemsToCart = (cartItems, cartItemToAdd) => {
  const existingCartItem = cartItems.find(cartItem => cartItem.id === cartItemToAdd.id)
  if(existingCartItem) {
    return cartItems.map(cartItem => 
        cartItem.id === cartItemToAdd.id ? 
          { ...cartItem, quantity: cartItem.quantity + 1} 
          : cartItem
       );
  }

  return [...cartItems, {...cartItemToAdd, quantity: 1}]
}

export const removeItemsFromCart = (cartItems, cartItemToDelete) => {
  return cartItems.filter(cartItem => cartItem.id !== cartItemToDelete.id );
}

export const removeSingleItemFromCart = (cartItems, cartItemToDelete) => {
  const existingCartItem = cartItems.find(cartItem => cartItem.id === cartItemToDelete.id)

  if(existingCartItem.quantity === 1) { 
    return cartItems.filter(cartItem => cartItem.id !== cartItemToDelete.id );
  }

  return cartItems.map(cartItem => 
    cartItem.id === cartItemToDelete.id ? 
      { ...cartItem, quantity: cartItem.quantity - 1} 
      : cartItem
   );
}