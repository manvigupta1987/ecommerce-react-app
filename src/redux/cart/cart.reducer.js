
// action is an object that contains two properties such as type and payload. Every reducer gets all the action fired by the view even if that 
// action is not related to that reducer. 

import CartActionType from './cart.types'
import {addItemsToCart, removeItemsFromCart} from './cart.utils'
const INTIAL_STATE = {
  hidden: true,
  cartItems : []
}
const cartReducer = (state=INTIAL_STATE, action) => {
  switch(action.type) {
    case CartActionType.TOGGLE_CART_HIDDEN:
      return {
        ...state,
        hidden: !state.hidden
      }
    case CartActionType.ADD_ITEM:
      return {
        ...state,
        cartItems: addItemsToCart(state.cartItems, action.payload)
      }
    case CartActionType.REMOVE_ITEM:
      return {
        ...state,
        cartItems: removeItemsFromCart(state.cartItems, action.payload)
      }
    default:
      return state;
  }
}

export default cartReducer;