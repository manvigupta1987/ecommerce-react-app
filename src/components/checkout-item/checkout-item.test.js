import React from 'react';
import { shallow } from 'enzyme';

import { CheckoutItem } from './checkout-item.component.jsx'
import {removeItems, addItems, removeSingleItem} from '../../redux/cart/cart.action'

describe('CheckoutItem component', () => {
  let wrapper;
  let mockDispatch;
  const mockCartItem = [{ id: 1 }];

  beforeEach(() => {
    mockDispatch = jest.fn();
    const mockProps = {
      cartItem: mockCartItem,
      dispatch: mockDispatch
    };

    wrapper = shallow(<CheckoutItem {...mockProps} />);
  });

  it('should render CheckoutItem component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should call dispatch with removeSingleItem when button is clicked', () => {
    let payload = [{id:1}]
    wrapper.find('.arrow').at(0).simulate('click');
    expect(mockDispatch).toHaveBeenCalledWith(removeSingleItem(payload));
  });

  it('should call dispatch with addItems when button is clicked', () => {
    let payload = [{id:1}]
    wrapper.find('.arrow').at(1).simulate('click');
    expect(mockDispatch).toHaveBeenCalledWith(addItems(payload));
  });

  it('should call dispatch with removeItems when button is clicked', () => {
    let payload = [{id:1}]
    wrapper.find('.remove-button').simulate('click');
    expect(mockDispatch).toHaveBeenCalledWith(removeItems(payload));
  });
});