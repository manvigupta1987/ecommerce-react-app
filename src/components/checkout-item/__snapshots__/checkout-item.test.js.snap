// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckoutItem component should render CheckoutItem component 1`] = `
<div
  className="checkout-item"
>
  <div
    className="image-container"
  >
    <img
      alt="product item"
    />
  </div>
  <span
    className="name"
  />
  <span
    className="quantity"
  >
    <div
      className="arrow"
      onClick={[Function]}
    >
      ❮
    </div>
    <span
      className="value"
    />
    <div
      className="arrow"
      onClick={[Function]}
    >
      ❯
    </div>
  </span>
  <span
    className="price"
  />
  <div
    className="remove-button"
    onClick={[Function]}
  >
    ✕
  </div>
</div>
`;
